[Jung] 미션5: 보드에 위치 부여 및 점수 계산 PR입니다.

rla36 opened the pull request:
안녕하세요? Jung입니다!
커밋 개수에서 드러나듯이,,, 미션5 하면서 고민이 정말 많았습니다.. 

###   주요 변화
- Color, Type 이넘 Piece클래스 안으로 이동
- Rank 클래스 생성
- Position 클래스 생성

### 고민한 것들
**Rank클래스**
- 기존에 Board에서 Piece를 직접 담당했었는데, 이것을 Rank클래스로 대부분 넘겨버렸습니다. Rank가 Piece를 담당하고, Board는 Rank를 담당하도록 역할 분배했습니다.
- 이렇게 역할 분배를 한 결과, Board에서 Piece를 직접 접근하기 위해선 for문을 2개 중첩해 사용해야 하는 문제가 발생했습니다. 가독성이 떨어져 어떻게 고칠지 고민했는데 별 성과는 없었습니다.. 조언부탁드립니다!

**Board클래스**
- Board의 calculatePoints메서드에서 pawn이 같은 file에 있으면 개다 0.5점씩 차감을 합니다. Board에서는 Rank를 다루고, Rank에서는 Piece를 다루기에 Board에서 직접 file에 접근할 수 있는 방법이 없어 코드가 복잡해졌습니다. 더 나은 해결책이 있을 것 같은데 고민할 시간이 필요할 것 같습니다..

**Position 클래스**
- Position 인스턴스는 같은 rank, file이라면 같은 위치로 생각해도 되므로 VO에 해당된다고 생각해 equals, hashCode를 오버라이딩했습니다.
- Position에서 생성자를 private으로 설정하고 Position.at("a4") 와 같이 at메서드를 통해 인스턴스를 받도록 했습니다. 이렇게 하면 인스턴스의 무분별한 생성을 막고, 메모리 과부하를 방지할 수 있다는 이야기를 들었습니다. [출처](https://stackoverflow.com/questions/12901734/what-is-the-advantage-of-singleton-design-pattern) 
- Board의 첫번쨰 고민과 관련해서, Board에서 file을 직접 접근할 수 없으므로, 차라리 Position안에 file이넘, rank이넘을 만들면 어떨까 생각했습니다. 그렇지만 게임플레이어가 위치를 String으로 입력할 것이기에 Position의 생성자 매개변수로 String을 받는 이점이 분명하다고 생각했습니다. (아니면 at메서드로 String을 입력받아 분해해 이넘으로 만든 후에 생성자의 매개변수에 이넘 넣어 return new Position(이넘, 이넘)하는 방법도 있을 것 같습니다.. 휴..)

Board에서 Rank클래스를 분리하는 데서 대부분의 고민이 생긴것 같은데, Board가 비대해지더라도 Rank의 역할을 Board로 통합시키는 것이 어떨지 다시 한번 생각해봐야할 것 같습니다.

리뷰해주셔서 감사합니다!

rla36 committed:
Color, Type enum을 Piece 클래스로 통합

pieces 디렉토리 아래 있던 Color, Type enum을 Piece클래스 내부 이넘으로
바꾸었습니다. Color에 NO_COLOR을 추가했습니다.
Type에 NO_PIECE를 추가하고, 생성자에 소문자 하나만을 받도록했습니다.
Piece클래스의 getRepresentation()메서드를 통해 Color, Type에 적절한
char을 리턴하도록 했습니다.

rla36 committed:
Color 이넘에 getWhite/BlackRepresentation 메서드 추가

getWhiteRepresentation으로 representation을 반환하고,
getBlackRepresentation을 통해 representation을 대문자로 바꾼 char을
리턴하도록 했습니다.

rla36 committed:
test: Color의 getWhite/BlackRepresentation 테스트 추가

Color의getWhiteRepresentation, getBlackRepresentation 메서드를 검증하는
테스트를 추가했습니다.

rla36 committed:
NO_PIECE enum의 representation '*'으로 수정

rla36 committed:
feat: createBlack() 메서드 추가

rla36 committed:
test : createBlank로 만들어진 Piece의 속성 검증

createBlank()로 만들어진 Piece의 Color은 NO_COLOR, Type은 NO_PIECE,
representation은 * 입니다. 이것을 검증하는 테스트를 작성했습니다.

rla36 committed:
feat: createWhite(), createBlack()메서드 추가, 중복제거

createWhite(), createBlack()메서드를 private메서드로 추가해 팩토리 메서드의
중복을 제거했습니다.

rla36 committed:
Rank클래스 생성.

Rank클래스 생성해 기존의 whitePieceList, blackPieceList와 메서드들이
담당하던 기능을 Rank로 옮기도록 했습니다.

rla36 committed:
feat: Rank에 addWhite/BlackPiece 메서드 추가

기존 Board에 있던 메서드를 Rank로 옮겨 색깔이 적절하면 Rank의 piececList에
추가하도록 했습니다.

rla36 committed:
자동수정 오류 복원

name을 type으로 리네이밍하는 과정에서 잘못 포함된 오류 수정했습니다.

rla36 committed:
feat: getRepresentation, getPiece추가

Board에 있던 getRepresentation은 Rank에서 다루도록했습니다. 각 Rank에서
8피스씩 representation을 관리하고,
Board에선 각 Rank의 representation을 관리하는 것이 효율적, 역할이
명확하다고 생각한 것이 수정의 이유입니다.

rla36 committed:
test: size() 검증 테스트 작성

Rank클래스의 size메서드를 검증하는 테스트를 작성했습니다.
initialize된 rank들은 initialize로 8개의 피스를 갖고 있으므로 size()의
기댓값이 8이고, initializeBlankRank로 텅 빈 Rank는 기댓값이 0입니다.

rla36 committed:
feat: size()메서드, isPiece 메서드 추가

Rank의 pieceList에서 blank 아니고 piece 인것들만 세어 piece의 개수를
알수 있는 size() 메서드를 추가했습니다.
blank가 아니라 piece인지 판단해 piece이면 true를 리턴하는
isPiece메서드를 추가했습니다.

rla36 committed:
test: getRankRepresentation을 검증하는 테스트 작성

getRankRepresentation을 검증하는 테스트를 작성했습니다.
blankRank의 경우 공백표시인 ********가 나오는지 확인했습니다.

rla36 committed:
getRepresentation 수정, size()메서드 rankPieceSize로 리네이밍

기존에 size()메서드는 rank에 들어있는 piece의 개수를 리턴하도록 했는데,
blank도 piece로 인식되면서0개가 아닌 8개의 피스가 있다고 말하는 문제가
있었습니다. 이것을 수정했고, 혼동이 발생하지 않도록 rankPieceSize로
이름을 수정했습니다.
getRepresentation의 수정도 size()메서드 떄문에 혼동이 발생해
수정했습니다.

rla36 committed:
Rank로 이동한 메서드 및 변수 제거

Rank에서 관리하도록 이동된 메서드와 변수를 Board클래스에서 제거했습니다.

rla36 committed:
test: 수정된 Board의 메서드 검증

boardLayoutToString이 rank의 representation을 리턴하는지,
boardPieceSize가 board의 모든 piece개수를 리턴하는지 검증하는 테스트를
작성했고, 쓰이지 않는 테스트는 지웠습니다.

rla36 committed:
BoardTest에서만 쓰이는 Board의 변수를 BoardTest로 이동

더이상 Board에서 쓰이지 않는 변수를 BoardTest의 메서드 안 지역변수로
이동시켰습니다.

rla36 committed:
test: certainPieceSize() 검증 테스트 작성

특정한 타입과 색깔의 piece가 보드위에 몇개 있는지 알수있는
certainPieceSize 메서드를 검증하기 위한 테스트를 작성했습니다.

rla36 committed:
feat: certainPieceSize메서드 추가

특정 색상과 타입의 piece가 보드 위에 몇개 있는지 알려주는 메서드를
작성했습니다.

rla36 committed:
test: getRank()로 호출되는 rank 정확한지 검증 테스트

getRank(int index)로 Rank를 호출 시, getRank(8)로 호출시 가장 윗줄의
검은 피스들이 호출되는지, getRank(2)로 호출시 흰색pawn이 호출되는지
검증하는 테스트를 작성했다.

rla36 committed:
feat: getRank(int index) 메서드 추가

getRank로 Rank를 호출시 8을 호출하면 가장 윗줄의 검은 piece가 담긴
rank가  호출되도록 했습니다. 즉 Board의 rankList에 담긴 순서와 반대되는
순서로 rank다 리턴되도록 했습니다. 체스게임의 컨벤션에 맞춰 8rank가 가장
윗줄이 되도록하기 위해 해당 기능을 추가 했습니다.

rla36 committed:
변수 리네이밍

rla36 committed:
test: findPiece검증 테스트 작성

findPiece(String location)으로 주어진 위치에 놓인 말을 찾을 수 있는지
검증하는 테스트를 작성했습니다.

rla36 committed:
feat: Board에 findPiece()메서드 추가

findPiece() 메서드로 주어진 위치의 말을 리턴하는 메서드를 작성했습니다.
a1~h8사이의 적절한 위치가 주어지지 않으면 예외를 발생시키도록 했습니다.

rla36 committed:
feat: equals, hashCode 메서드 오버라이딩 추가

Piece끼리 비교할때 Color과 Type가 일치하면 같은 객체로 인식하도록
equals()와 hashCode를 오버라이딩했습니다. 이렇게 한 이유는 Board클래스의
findPiece()메서드를 사용했을떄 리턴되는 Piece가, createPiece로 만들어진
Piece와 일치하도록 하기 위해서입니다.

rla36 committed:
test: Piece의 equals, hashCode 검증 테스트 추가

equals, hashCode오버라이딩이 제대로 되어 같은 Color와 Type이면 동일한
객체로 인식하는지 확인하기 위해 테스트를 추가했습니다.

rla36 committed:
test: findPiece 검증 테스트 보완

rla36 committed:
test: Board의 initializeEmptyBoard 테스트 작성

initializeEmptyBoard로 Board에 아무것도 올려지지 않은 상태가 되는지
확인하는 테스트 작성했습니다.

rla36 committed:
feat: initializeEmptyBoard로 빈상태 초기화

initializeEmptyBoard로 아무것도 올려지지 않은 상태로 초기화되는 메서드를
작성했습니다.

rla36 committed:
Position클래스 추가

Position은 int로 변환된 fileIndex와 rankIndex를 인스턴스변수로 갖습니다.
생성자가 private하기 때문에 외부에서는 at(String str) 메서드를 통해 스트링으로 입력받은 값의 Position 인스턴스를 리턴하도록했습니다.

잘못된 범위의 위치를 입력한 경우는 at() 메서드에서 전부 처리하도록
했습니다.

rla36 committed:
Board의 위치 예외 제거

Position에서 위치의 모든 예외를 담당하도록 했기 때문에 제거했습니다.

rla36 committed:
feat: replace메서드 구현

Rank안의 fileIndex의 자리에 특정 piece를 넣는 메서드 구현했습니다.

rla36 committed:
test: Rank의 replace메서드 테스트 추가

rla36 committed:
test: Board의 addNewPiece 테스트 추가

addNewPiece로 특정 자리에 만들어진 piece가, Board의 findPiece로 위치를
입력했을때 발견되는 piece와 일치하는지 검증하는 테스트를 작성했습니다.

rla36 committed:
feat: addNewPiece 메서드 추가

Board의 특정 위치에 새로 생긴 piece를 추가하는 메서드를 추가했습니다.
미션 상에는 move로 되어있었으나, 원래 있던 piece를 옮기는 것이 아니라
새로 만들어진 piece를 특정 자리에 놓는 것이기 떄문에 메서드 명을
addNewPiece와 같이 했습니다.

rla36 committed:
isPiece메서드 Piece클래스로 이동

rla36 committed:
test: movePiece검증 테스트

movePiece로 한 곳에서 piece가 발견되면 다른 곳으로 옮기는지 검증하는
테스트를 작성했습니다.

rla36 committed:
feat: movePiece메서드 추가

movePiece(String from, String to)로 from 으로 얻은 position에서 piece가
존재할 시에, 그 piece를 to position으로 옮기는 메서드를 추가했습니다.

rla36 committed:
feat: calculatePoints메서드 추가

색깔별로 점수를 합산해 리턴하는 메서드를 추가했습니다. calculatePoints는
추가적으로 두개의 Board메서드를 사용하는데, getPieceListByColor과
pawnSizeInFile입니다. getPieceListByColor은 특정 색상의 piece들을 모두
arrayList에 담아 리턴합니다. pawnSizeInFile은 동일한 File에 특정 색상의
pawn이 두개 이상 있을 경우, 그런 폰들의 개수를 모두 합해 리턴합니다.
calculatePoints는 매개변수로 주어진 색상으로 getPieceListByColor메서드를
사용해 특정 색상의 모든 디폴트포인트를 합산하고, 그 후에 동일한 파일에
pawn이 존재한다면 0.5포인트씩 차감해 총 포인트를 계산합니다.

rla36 committed:
test: calculatePoints 검증 메서드 추가

rla36 committed:
feat: Piece클래스에 isColorType 메서드 추가

Piece클래스에 특정 색깔, 특정 타입이면 true를 리턴하는 메서드를 추가했습니다.
이 메서드를 사용해  Board클래스의  certainPieceSize 메서드에서 특정 색, 타입의
piece가 맞다면 사이즈를 1증가시키는 부분을 간결화했습니다.
`

rla36 committed:
feat: Piece의 equals와 hashCode 오버라이드

Piece가 동일한 파일, 랭크이면 동일한 값으로 인식되도록(VO) equals와
hashCode을 오버라이드했습니다.

rla36 committed:
test: Positon의 equals 검증 테스트 추가

rla36 committed:
feat: compareTo 오버라이드, Comparable인터페이스 구현

Piece끼리 비교할때, 점수를 기준으로 비교해 매개변수로 입력되는 piece보다
크면 1을, 작으면 -1을, 같으면 0을 리턴하도록 compareTo를
오버라이딩했습니다.
Comparable<Piece> 인터페이스를 구현하도록했습니다.

rla36 committed:
test: sortColorPieces Ascending/Descending 테스트 추가

Board에 추가된 sortColorPiecesAscending,
sortColorPiecesDescending메서드가 제대로 작동하는지 검증하는 테스트를
추가했습니다.

rla36 committed:
feat: Board에 sortColorPieces Ascending/Descending 추가

매개변수로 주어진 색상의 piece들을 getPieceListByColor메서드를 통해
List에 담아, sortColorPiecesAscending/Descending으로 오름차순,
내림차순으로 정렬하도록했습니다.

rla36 committed:
쓰이지 않는 메서드 제거(Rank의 resetRank)

rla36 committed:
재정렬, 이름 수정, 중복제거

sonypark commented on src/test/java/net/jung/chess/pieces/PieceTest.java:
```diff
@@ -50,4 +52,40 @@ void checkIsColor(){
+        );
+    }
+
+    @Test
+    @DisplayName("createBlack로 만들어진 Piece는 Color.NO_COLOR, Type.NO_PIECE이다. ")
```
DisplayName 오타! -> checkCreateBlank로 변경
처음에 보고 'Black이 왜 NO Color이지?' 했네요🤔

rla36 replied:
앗 실수.. 꼼꼼하게 봐주셔서 감사합니다!! 수정하겠습니다 ㅎㅎ

sonypark commented on src/main/java/net/jung/chess/Position.java:
```diff
@@ -0,0 +1,59 @@
+
+    public static Position at(String location) throws IllegalArgumentException {
+        char[] positionChar = location.toCharArray();
+        if (!(location.length() == 2
+                && isValidFile(positionChar[0])
+                && isValidRank(positionChar[1]))) {
+            throw new IllegalArgumentException("a1~h8 범위의 위치를 입력해주세요.");
+        }
```
예외 처리 👍
2와 같은 매직 넘버는 피해주세요~

rla36 replied:
네 2를 Postion클래스에 private static final int NUM_OF_DIGITS = 2 를 추가해 2를 대체하도록 수정했습니다.!

sonypark commented on src/main/java/net/jung/chess/Position.java:
```diff
@@ -0,0 +1,59 @@
+package net.jung.chess;
+
+public class Position {
+    private int fileIndex;
+    private int rankIndex;
```
final을 붙여도 좋겠네요

rla36 replied:
한번생성되면 바뀔일이없으니까 그렇겠네요! 변수, 메서드 속성 정하는게 아직도 많이 부족합니다 ㅎ.. 감사합니다!!

sonypark commented on src/main/java/net/jung/chess/Piece.java:
```diff
@@ -0,0 +1,113 @@
+        this.color = color;
+    }
+
+    public Color getColor() { return color; }
+
+    public Type getType() { return type; }
+
+    public char getRepresentation() {
+        return color == Color.WHITE ? type.getWhiteRepresentation() : type.getBlackRepresentation();
+    }
+
+    public double getDefaultPoint() {
+        return type.defaultPoint;
+    }
```
getter 메서드는 하단에 배치해주세요. (Overriding 한 메서드 위)

rla36 replied:
메서드 위치에도 컨벤션이 있었군요..! 수정했습니다! 컨벤션을 공부해보겠습니다

sonypark commented on src/main/java/net/jung/chess/Board.java:
```diff
@@ -1,132 +1,142 @@
+    public int boardPieceSize() {
+        int boardPieceSize = 0;
 
-    public void addWhitePiece(Piece piece) {
-        if (piece.isWhite()) {
-            whitePieceList.add(piece);
+        for (Rank rank : rankList) {
+            boardPieceSize += rank.rankPieceSize();
         }
+        return boardPieceSize;
     }
```
```java
    public int boardPieceSize() {
        return rankList.stream()
                .mapToInt(Rank::rankPieceSize)
                .sum();
    }
```
(참고) 스트림을 이용하면 이렇게 구현할 수도 있습니다.

rla36 replied:
감사합니다 부끄럽지만 제가 아직 스트림 공부를 못해서..ㅜ 일단 말씀해주신대로 수정하고 공부해보겠습니다 !!!

sonypark commented on src/main/java/net/jung/chess/Board.java:
```diff
@@ -1,132 +1,142 @@
 
-    public int blackPieceListSize() {
-        return blackPieceList.size();
+    public String boardLayoutToString() {
+        StringBuilder boardLayout = new StringBuilder();
+        for (Rank rank : rankList) {
+            boardLayout.append(appendNewLine(rank.rankLayoutToString()));
+        }
+        return boardLayout.toString();
     }
```
화면에 표현해주기 위한 로직은 이곳에서 하는게 아니라 Main 또는 출력을 담당하는 클래스에서 해주는게 좋을 것 같아요.
Board는 체스판만 리턴하고 출력은 외부에서 해주는 방식

rla36 replied:
넵 조금더 고민해보겠습니다...!

sonypark commented on src/main/java/net/jung/chess/Rank.java:
```diff
@@ -0,0 +1,106 @@
+        return pieceList.get(index);
+    }
+
+    public String rankLayoutToString() {
+        StringBuilder strBuilder = new StringBuilder();
+        for (int i = 0; i < pieceList.size(); i++) {
+            strBuilder.append(getPiece(i).getRepresentation());
+        }
+        return strBuilder.toString();
+    }
```
여기도 마찬가지로, rank 는 pieces만 리턴하고 출력을 위한 로직은 외부에서 하는 방식을 생각해보세요

sonypark commented on src/main/java/net/jung/chess/Rank.java:
```diff
@@ -0,0 +1,106 @@
+
+public class Rank {
+    public static final int MAX_PIECES_IN_RANK = 8;
+    private ArrayList<Piece> pieceList = new ArrayList<>();
```
변수 타입은 ArrayList보단 List로 하는게 좋을 것 같습니다. 왜 그럴까요?
List와 같은 자료형의 이름은 변수명에 넣지 않는걸 권장합니다.
pieceList보단 pieces가 낫겠네요

rla36 replied:
앗 다른 곳에서는 List로 했는데 왜 저기만 ArrayList가 들어갔을까요.. 다형성 때문에 List로 쓰는 경우가 많다고 알고 있습니다! 말씀하신 사항 수정했습니다~ (Board의 rankList도 ranks로 수정했습니다!)

sonypark commented on src/main/java/net/jung/chess/Board.java:
```diff
@@ -1,132 +1,142 @@
+        double points = 0;
+
+        List<Piece> colorPieceList = getPieceListByColor(color);
+        for (int i = 0; i < colorPieceList.size(); i++) {
+            points += colorPieceList.get(i).getDefaultPoint();
+        }
```
```suggestion
        points = colorPieceList.stream()
                .mapToDouble(Piece::getDefaultPoint)
                .sum();
```
여기도 stream으로 바꾸거나 enhanced for loop를 사용할 수 있겠네요

rla36 replied:
여기서도 마찬가지로 일단 수정한 후에 스트림을 제대로 공부하겠습니다!

sonypark commented on src/main/java/net/jung/chess/Board.java:
```diff
@@ -1,132 +1,142 @@
 
+    public List<Piece> sortColorPiecesDescending(Piece.Color color) {
+        List<Piece> colorPieceList = getPieceListByColor(color);
+        Collections.sort(colorPieceList, Collections.reverseOrder());
```
```suggestion
        colorPieceList.sort(Collections.reverseOrder());
```
List에도 sort라는 내장 함수가 있습니다.

sonypark reviewed (changes requested):
안녕하세요 Jung, 리뷰어 소니 입니다.😊
구현한 코드 잘 보았습니다.
특히 테스트 코드를 꼼꼼히 짜려고 노력하신 것 같아 좋았어요!
커밋 로그를 봐도 고민 많이 한 흔적이 보이네요.
간단한 코멘트 남겼으니 확인 부탁드립니다~

sonypark commented on src/main/java/net/jung/chess/Position.java:
```diff
@@ -0,0 +1,59 @@
+        fileIndex = location.toCharArray()[0] - 'a';
+        rankIndex = location.toCharArray()[1] - '0';
+    }
+
+    public static Position at(String location) throws IllegalArgumentException {
```
at 보다는 정적 팩토리 메서드에서 주로 사용하는 네이밍을 사용하시면 더 좋을 것 같네요.
어떤 네이밍을 주로 사용하는지 찾아보세요~

rla36 replied:
factory method naming convention검색하니 create를 많이 쓴다고 해 createPosition으로 수정하도록 하겠습니다!
Position.at("a6")식으로 하면 의미가 직관적으로 와닿는다고 생각했는데 컨벤션이 중요하단 것을 몰랐습니다ㅜ 알려주셔서 감사합니다

honux77 replied:
https://stackoverflow.com/questions/3368830/how-to-name-factory-like-methods 요 링크도 참고하세요.
제 생각에는 of 괜찮을 것 같은데 어때요?

rla36 committed:
Position의 인스턴스 변수  final 설정, 리네이밍

position클래스에 static변수 NUM_OF_DIGITS 추가

rla36 committed:
getter메서드 재배치

rla36 committed:
Board클래스 boardPieceSize메서드 스트림으로 수정

rla36 committed:
변수명 리네이밍, ArrayList형 변수  List로 선언

Rank의 pieceList의 이름을 pieces로 리네이밍하고, 선언부에서 List로
선언되도록 수정했습니다. 추가적으로 Board의 rankList도 ranks로
수정했습니다.

rla36 committed:
Board클래스 calculatePoints메서드 수정

for문 스트림을 사용하도록 수정했습니다

Co-authored-by: Kong Son Park <sonypark0204@gmail.com>

rla36 committed:
sortColorPiecesDescending 메서드 수정

List의 sort사용하도록 수정했습니다.

Co-authored-by: Kong Son Park <sonypark0204@gmail.com>

rla36 committed:
Position 팩토리 메서드 리네이밍

at메서드를 createPosition으로 리네이밍했습니다.

rla36 committed:
Merge branch 'mission5' of https://github.com/rla36/java-chess into mission5

sonypark reviewed (approved):
수정 사항 확인했습니다.
고생 많으셨습니다!
